sympy/solvers/solvers.py

```diff
@@ -995,26 +997,87 @@ def _sympified_list(w):
+            ri = a.args[0].as_real_imag()
+            if any(_.has(*symbols) for _ in ri for _ in  _.atoms(re, im)):
                 raise NotImplementedError('solving %s when the argument '
-                    'is not real or imaginary.' % e)
+                'is not real or imaginary.' % a)
+            if 0 not in ri:
+                # if real, ri will be (a.args[0], 0)
+                # if imaginary, ri will be (0,I*a.args[0])
+                raise NotImplementedError('solving %s with real and '
+                'imag parts.' % a)
+            abs_[a] = sum(ri)
+
+    if abs_:
+        abs_nodes, args = zip(*abs_.items())
+        abs_nodes = list(ordered(abs_nodes))
+        # identify the outermost abs that appear
+        outer = []
+        for i in range(len(abs_nodes)):
+            if not any(abs_nodes[j].has(abs_nodes[i]) for j in
+                    range(i + 1, len(abs_nodes))):
+                outer.append(abs_nodes[i])
+        # now figure out what values they can have while considering
+        # any nested abs within them
+        absset = set(abs_nodes)
+        del abs_nodes
+        signed = []
+        for o in outer:
+            a = list(reversed(list(ordered([i for i in o.atoms(Abs)
+                if i in absset]))))
+            b = [abs_[a.pop(0)]]
+            b.append(-b[0])
+            for i in a:
+                b = [j.subs(i, abs_[i]) for j in b
+                    ] + [j.subs(i, -abs_[i]) for j in b]
+            signed.append(b)
+        # capture and set flags
+        as_dict = flags.get('dict', None)
+        flags['dict'] = True  # to make work here easier
+        check = flags.get('check', True)
+        flags['check'] = False  # we need to check in original f
+        # now start finding possible solutions
+        sol = []
+        linear = True
+        for v in product(*signed):
+            reps = dict(zip(outer, v))
+            fi = [_.xreplace(reps) for _ in f]
+            _linear, s = _get_solution(fi, bare_f, symbols, flags)
+            if linear:
+                # it is possible that a nonlinear system
+                # will appear linear, e.g. the square terms
+                # can cancel in abs(x**2 - 1) - x**2 + x - 4
+                linear = _linear
+            sol.extend(s)
+
+        # remove redundant solutions before doing any checking
+        # since this is a costly step
+        solution = sol = list(uniq(sol))
 
-        # arg
-        fi = fi.replace(arg, lambda a: arg(a).rewrite(atan2).rewrite(atan))
+        if check:
+            # check assumptions
+            sol = _check_assumptions(sol, flags.get('warn', False))
+            # need to check in the original equation
+            solution = []
+            for si in sol:
+                if checksol(f, si) is not False:
+                    solution.append(si)
 
-        # save changes
-        f[i] = fi
+        return _legacy_output(solution, symbols, ordered_symbols, bare_f, linear, as_dict, as_set)
```

Comment:
The fact that much of this code is duplicated from below makes me think that this is all being handled in the wrong place. It should probably be handled inside `_solve` or `_solve_system` rather than here.

Factoring out code like `_legacy_output` is good but we shouldn't be duplicating the return logic here anyway.

Reply:
> we shouldn't be duplicating the return logic here anyway.

We have to duplicate the logic everywhere we return choose to return a solution. This PR adds a new return location.

> this is all being handled in the wrong place

Checking of solutions (not assumptions) and simplification is normally handled in the `_solve*` routines. In this PR, however, that checking of solutions is shut off because it would be redundant: we need to check the solutions in the original set of equations. Only the calling routine knows what the original equations are. If you think it would be better,  a `_solve_abs` could be extracted that contains this logic.

Reply:
> We have to duplicate the logic everywhere we return choose to return a solution. This PR adds a new return location.

Adding a new return location is precisely part of the problem though. This function is a big mess and the more control paths we add the more complicated it becomes.

> we need to check the solutions in the original set of equations.

This is something that should be handled better in general. Note that these are already not the original equations which have already modified before we reach this point.

More generally the checking is all handled badly anyway. We should not check by substituting into the equations but rather by keeping track of the conditions that actually need to be checked. See `cond` in my suggested approach above and note that the conditions to be checked are *inequality* conditions that are easily checked through numerical evaluation rather than equality conditions that are impossible in general to verify.

In the symbolic case we cannot use numerical evaluation but then the validity itself is uncheckable anyway:
```python
In [11]: x, y = symbols('x, y',real=True)

In [12]: solve([abs(x) - 1], [x])
Out[12]: [(-1,), (1,)]

In [13]: solve([abs(x) + 1], [x])
Out[13]: []

In [14]: solve([abs(x) + y], [x]) # What should this return?
Out[14]: []
```

Reply:
> ```python
> 
> In [14]: solve([abs(x) + y], [x]) # What should this return?
> ```

Since `x` must be real, in this simple case we know that `y` must be real and, if it is, then `y + y` is not 0 so there is no solution. The previous version of this PR gave that result.

> More generally the checking is all handled badly anyway. We should not check by substituting into the equations but rather by keeping track of the conditions that actually need to be checked.

Reply:
y could be -1